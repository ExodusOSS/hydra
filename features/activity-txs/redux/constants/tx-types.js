export const TX_TYPES = {
  EXCHANGE: 'exchange',
  FIAT: 'fiat',
  RECEIVED: 'rx',
  SENT: 'tx',
  APPROVAL: 'approval',
  CLEANUP: 'cleanup',
  NFT: 'nft',
}
