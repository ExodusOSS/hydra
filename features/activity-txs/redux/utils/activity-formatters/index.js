import { formatSwapActivity } from './format-swap-activity'
import { formatTxActivity } from './format-tx-activity'

export const formattersByType = new Map([
  ['exchange', formatSwapActivity],
  ['tx', formatTxActivity],
])
