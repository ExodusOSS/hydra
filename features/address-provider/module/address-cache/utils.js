import lodash from 'lodash'
import assert from 'minimalistic-assert'
import { set } from '@exodus/basic-utils'

const { get } = lodash

// This allows new assets to use a full key, without regenerating existing data in fusion for legacy ones. Please do NOT add new names here!
const legacyKeyAssetNames = new Set([
  'algorand',
  'ark',
  'avalanchec',
  'bcash',
  'bgold',
  'bitcoin',
  'bitcoinsv',
  'bnbmainnet',
  'bsc',
  'cardano',
  'cosmos',
  'dash',
  'decred',
  'digibyte',
  'dogecoin',
  'elrond',
  'eosio',
  'ethereum',
  'ethereumarbnova',
  'ethereumarbone',
  'ethereumclassic',
  'ethereumgoerli',
  'fantommainnet',
  'filecoin',
  'flare',
  'harmonymainnet',
  'hedera',
  'iconmainnet',
  'lightningnetwork',
  'lisk',
  'litecoin',
  'matic',
  'neo3',
  'monero',
  'nano',
  'nem',
  'neo',
  'ontology',
  'optimism',
  'polkadot',
  'qtumignition',
  'ravencoin',
  'ripple',
  'solana',
  'stellar',
  'terra',
  'tezos',
  'theta',
  'tronmainnet',
  'vechainthor',
  'vertcoin',
  'waves',
  'zcash',
  'zilliqa',
])

export const getCachePath = ({ walletAccountName, baseAssetName, derivationPath }) => {
  assert(typeof walletAccountName === 'string', 'expected string "walletAccountName"')
  assert(typeof baseAssetName === 'string', 'expected string "baseAssetName"')
  assert(typeof derivationPath === 'string', 'expected string "derivationPath"')

  // todo add assetName for all assets when we migrate to a new fusion channel
  let path = derivationPath
  if (!legacyKeyAssetNames.has(baseAssetName)) {
    path = `${baseAssetName}/${path}` // combine these two to match fusion shape
  }

  return [
    // walletAccount is indirectly captured in the BIP path but it is still required because the BIP path is not enough, Trezor's accountIndex always use 1 regardless of the index of a walletAccount.
    walletAccountName,
    // assetName is required because some assets encode the network identifier into the address e.g splitting mainnet vs testnet results in different addresses for the same public key.
    // bipPath is required for assets that may have multiple addresses, or stake addresses etc
    path,
  ]
}

export const diffCaches = (cache1, cache2, fromSync) => {
  let isDifferent = false
  let needsSync = false

  const result = Object.create(null)
  for (const walletAccount in cache2) {
    const subCache = cache2[walletAccount]
    for (const path in subCache) {
      const existing = get(cache1, [walletAccount, path])
      const incoming = subCache[path]
      const newAddress = !existing
      const addressMismatch = !newAddress && existing.address !== incoming.address
      needsSync = needsSync || existing?.synced === false
      const justSynced = !existing?.synced && incoming.synced
      if (fromSync && addressMismatch && !existing?.synced) {
        // We have an address scheduled to sync up, that is mismatching with a
        // sync down, in this case we actually prefer the one scheduled to sync up.
        set(result, [walletAccount, path], existing)
      } else if (newAddress || addressMismatch || justSynced) {
        isDifferent = true
        set(result, [walletAccount, path], incoming)
      }
    }
  }

  return {
    diff: result,
    isDifferent,
    needsSync: isDifferent || needsSync,
  }
}

export const getUnsyncedAddressesForPush = (cache, currentCache) => {
  const out = Object.create(null)
  for (const walletAccount in cache) {
    const subCache = cache[walletAccount]
    for (const path in subCache) {
      const item = subCache[path]
      const currentCacheItem = get(currentCache, [walletAccount, path])
      if (
        currentCacheItem &&
        currentCacheItem.synced &&
        currentCacheItem.address.toString() === item.address.toString()
      )
        continue
      if (!item.synced) set(out, [walletAccount, path], item.address)
    }
  }

  return out
}
