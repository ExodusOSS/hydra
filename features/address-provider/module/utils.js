import { Address, AddressSet } from '@exodus/models'
import lodash from 'lodash'
import assert from 'minimalistic-assert'
import BipPath from 'bip32-path'
import { isNil } from '@exodus/basic-utils'
import KeyIdentifier from '@exodus/key-identifier'

const { groupBy } = lodash

export const resolvePurpose = ({ asset, address }) => {
  // Ideally, We should Unify the creation of Address object to always set the purpose
  if (address.meta.purpose !== undefined) {
    return address.meta.purpose
  }

  // TODO: store address.meta.purpose directly
  const purposeFromIndividualFlags = address.meta.isSegwit
    ? 84
    : address.meta.isNestedSegwit
      ? 49
      : address.meta.isTaproot
        ? 86
        : null

  if (purposeFromIndividualFlags !== null) {
    return purposeFromIndividualFlags
  }

  const addressString = address.toString()
  if (asset.address?.resolvePurpose) {
    return asset.address.resolvePurpose(addressString) || null
  }

  // fallback when resolvePurpose is not implemented (a default implementation could be like below though)
  if (asset.address?.validate) {
    try {
      return asset.address.validate(addressString) ? 44 : null
    } catch (e) {
      // some assets' validate may raise an error when invalid. imho, this should be fixed.
      console.warn(`${asset.name} asset.address.validate function raises an error!!!`, e)
    }
  }

  return null
}

export const getAddressesFromTxLog = ({ asset, txLog, walletAccount }) => {
  const changeAddresses = AddressSet.fromArray(
    [...txLog].map((tx) => tx.data?.changeAddress).filter(Boolean)
  )
    .toArray()
    .filter((address) => address.pathArray[0] === 1) // changeAddress may contain receive address

  const allAddresses = txLog.addresses.union(changeAddresses).sort(AddressSet.PATH_SORTER)

  return [...allAddresses].map((address) => {
    const purpose = resolvePurpose({ asset, address })

    const chainIndex = address.pathArray[0]
    const addressIndex = address.pathArray[1]
    return {
      address: new Address(address.toString(), {
        ...address.meta,
        purpose,
        walletAccount: walletAccount.toString(),
        keyIdentifier: new KeyIdentifier(
          asset.baseAsset.api.getKeyIdentifier({
            purpose,
            accountIndex: walletAccount.index,
            chainIndex,
            addressIndex,
            compatibilityMode: walletAccount.compatibilityMode,
          })
        ),
      }),
      purpose,
      chainIndex,
      addressIndex,
    }
  })
}

export const resolveUnusedAddressIndexesFromAddresses = ({
  addresses,
  purposes,
  highestUnusedIndexes,
}) => {
  const groupByPurpose = groupBy(addresses, 'purpose')
  const knownChains = [0, 1]
  return purposes.map((purpose) => ({
    purpose,
    chain: knownChains.map((thisChainIndex) =>
      (groupByPurpose[purpose] || [])
        .filter(({ chainIndex }) => thisChainIndex === chainIndex)
        .reduce(
          (acc, { addressIndex }) =>
            highestUnusedIndexes
              ? Math.max(addressIndex + 1, acc)
              : addressIndex === acc
                ? acc + 1
                : acc,
          0
        )
    ),
  }))
}

export const inclusiveRange = (end) => {
  return [...Array.from({ length: end }).keys(), end]
}

export const createPath = ({ chainIndex, addressIndex }) => {
  assert(
    !isNil(chainIndex) || isNil(addressIndex),
    'chainIndex must be provided if addressIndex is provided'
  )
  return BipPath.fromString(
    [chainIndex, addressIndex].filter((i) => !isNil(i)).join('/')
  ).toString()
}
