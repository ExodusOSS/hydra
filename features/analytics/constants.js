// countries for which we collect telemetry and offer opt-out
export const TELEMETRY_OPT_OUT_COUNTRIES = new Set([
  'US', // united states
  'AU', // australia
  'IN', // india
  'MX', // mexico
  'NZ', // new zealand
  'NG', // nigeria
  'ZA', // south africa
  'CH', // switzerland
])
