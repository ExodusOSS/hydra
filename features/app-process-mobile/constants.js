import { AppState } from 'react-native'

const startTime = Date.now()

const initialAppState = AppState.currentState || 'active'

export const APP_PROCESS_INITIAL_STATE = {
  mode: initialAppState,
  networkType: 'unknown',
  timeLastBackgrounded: initialAppState === 'background' ? startTime : 0,
  timeInBackground: 0,
  startTime,
}
