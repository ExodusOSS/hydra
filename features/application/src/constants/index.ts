export * from './lifecycle-hooks.js'
