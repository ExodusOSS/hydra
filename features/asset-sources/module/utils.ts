import typeforce from '@exodus/typeforce'
import { getTrezorMeta } from './trezor.js'
import assert from 'minimalistic-assert'
import type { Asset } from '../types.js'
import type { WalletAccount } from '@exodus/models'
import { types } from './types.js'

type GetSupportedPurposesOpts = {
  asset: Asset
  walletAccount: WalletAccount
}

export const getSupportedPurposes = (opts: GetSupportedPurposesOpts) => {
  typeforce(
    {
      asset: types.asset,
      walletAccount: types.walletAccount,
    },
    opts,
    true
  )

  const {
    asset,
    walletAccount: { compatibilityMode, isMultisig },
  } = opts

  const baseAsset = asset.baseAsset
  const apiExists = !!baseAsset.api?.getSupportedPurposes

  assert(
    !isMultisig || apiExists,
    'baseAsset must have api.getSupportedPurposes for a multisig WalletAccount'
  )

  if (apiExists) {
    // we seem to be having two different APIs for getSupportedPurposes, one that expects compatibilityMode separately, and one that expects it on walletAccount
    const supportedPurposes = baseAsset.api?.getSupportedPurposes!({
      ...opts,
      compatibilityMode,
      isMultisig,
    })
    assert(supportedPurposes?.length, 'at least one purpose must be supported')
    return supportedPurposes
  }

  // the given asset should know what's the best value depending on the wallet account type and the compatibility mode.
  // everything below here to be removed

  if (compatibilityMode === 'trezor') {
    const trezorMeta = getTrezorMeta(baseAsset)
    return [
      baseAsset.useBip84 && trezorMeta.isBip84(asset.name) && 84, // segwit
      trezorMeta.isSegwit(asset.name) ? 49 : 44, // nested segwit
    ].filter(Boolean)
  }

  return [
    baseAsset.useBip49 === true && 49, // nested segwit
    baseAsset.useBip84 === true && 84, // segwit
    44, // legacy
    baseAsset.useBip86 === true && 86, // taproot
  ].filter(Boolean)
}

export const getDefaultPurpose = (opts: GetSupportedPurposesOpts) => {
  return getSupportedPurposes(opts)[0]
}
