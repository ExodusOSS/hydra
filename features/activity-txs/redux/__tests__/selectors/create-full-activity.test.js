import { BUY_ORDER } from '@exodus/fiat-ramp/redux/__tests__/fixture'
import { FiatOrderSet, Tx } from '@exodus/models'
import { NFTS_NETWORK_TO_ASSET_NAME } from '@exodus/nfts/constants'

import { setup } from '../utils'

describe('createFullActivity', () => {
  test('should return empty array of activities in wallet', () => {
    const { store, selectors } = setup()
    const createSelector = selectors.activityTxs.createFullActivity({
      nftsNetworkNameToAssetName: NFTS_NETWORK_TO_ASSET_NAME,
    })

    const selector = createSelector({
      assetName: 'ethereum',
      walletAccount: 'exodus_0',
    })

    expect(selector(store.getState())).toEqual([])
  })

  test('should add nfts, fiat, connections to props', () => {
    const { store, selectors, handleEvent, assets } = setup()

    const txLogFixtures = {
      ethereum: [
        {
          txId: 'received-tx',
          error: null,
          date: '2019-07-22T13:54:28.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '1 ETH',
          coinName: 'ethereum',
          feeCoinName: 'ethereum',
          feeAmount: '0.000189 ETH',
          currencies: { ethereum: assets.ethereum.currency },
        },
        {
          txId: 'buy-order-tx',
          error: null,
          date: '2019-07-22T13:54:28.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '1 ETH',
          coinName: 'ethereum',
          feeCoinName: 'ethereum',
          feeAmount: '0.000189 ETH',
          currencies: { ethereum: assets.ethereum.currency },
        },
      ],
    }

    const nfts = {
      exodus_0: {
        ethereum: [
          {
            nftId: 'someId1',
            date: 1,
            from: 'ethereumAddress1',
            to: 'ethereumAddress2',
            txId: 'received-tx',
          },
        ],
      },
    }

    handleEvent('activityTxs', {
      exodus_0: {
        ethereum: txLogFixtures.ethereum.map((tx) => Tx.fromJSON(tx)),
      },
    })
    handleEvent('personalNotes', [
      { txId: 'received-tx', message: 'message', providerData: { origin: 'someOrigin' } },
    ])
    handleEvent('connectedOrigins', [{ origin: 'someOrigin', foo: 'bar' }])
    handleEvent('nftsTxs', nfts)
    handleEvent('fiatOrders', FiatOrderSet.fromArray([{ ...BUY_ORDER, toAsset: 'ethereum' }]))

    const createSelector = selectors.activityTxs.createFullActivity({
      nftsNetworkNameToAssetName: NFTS_NETWORK_TO_ASSET_NAME,
    })

    const selector = createSelector({
      assetName: 'ethereum',
      walletAccount: 'exodus_0',
    })

    const result = selector(store.getState())

    expect(result.length).toEqual(2)
    expect(result[0].id).toEqual('nft.ethereum.received-tx')
    expect(result[0].type).toEqual('nft')
    expect(result[0].personalNote).toEqual({
      txId: 'received-tx',
      message: 'message',
      providerData: { origin: 'someOrigin' },
    })
    expect(result[0].connection).toEqual({ origin: 'someOrigin', foo: 'bar' })
    expect(result[0].nft).toEqual({
      date: 1,
      from: 'ethereumAddress1',
      network: 'ethereum',
      nftId: 'someId1',
      to: 'ethereumAddress2',
      txId: 'received-tx',
    })

    expect(result[1].id).toEqual('ethereum.buy-order-tx')
    expect(result[1].fiatOrder.toJSON()).toEqual({
      date: '2023-07-11T15:10:42.226Z',
      exodusRate: 35,
      fiatValue: 22,
      fromAddress: null,
      fromAmount: 42,
      fromAsset: 'USD',
      fromWalletAccount: null,
      orderId: 'buy-order',
      orderType: 'buy',
      provider: 'ramp',
      providerRate: 32,
      status: '',
      toAddress: 'toAddress',
      toAmount: 42,
      toAsset: 'ethereum',
      toWalletAccount: 'exodus_0',
      txId: 'buy-order-tx',
    })

    expect(result).toMatchSnapshot()
  })

  test('should support limit prop', () => {
    const { store, selectors, handleEvent, assets } = setup()
    const fixtures = {
      ethereum: [
        {
          txId: 'tx-id-1',
          error: null,
          date: '2019-07-22T13:54:28.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '-0.23637286 ETH',
          coinName: 'ethereum',
          feeCoinName: 'ethereum',
          feeAmount: '0.000189 ETH',
          to: '0xa96b536eef496e21f5432fd258b6f78cf3673f74',
          currencies: { ethereum: assets.ethereum.currency },
        },
        {
          txId: 'tx-id-2',
          error: null,
          date: '2019-07-22T14:18:09.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '0.24511145 ETH',
          coinName: 'ethereum',
          currencies: { ethereum: assets.ethereum.currency },
        },
        {
          txId: 'tx-id-3',
          error: null,
          date: '2019-07-23T11:23:39.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '0.31622234 ETH',
          coinName: 'ethereum',
          currencies: { ethereum: assets.ethereum.currency },
        },
      ],
    }

    handleEvent('activityTxs', {
      exodus_0: {
        ethereum: fixtures.ethereum.map((tx) => Tx.fromJSON(tx)),
      },
    })
    handleEvent('personalNotes', [])
    handleEvent('connectedOrigins', [])
    handleEvent('nftsTxs', {})
    handleEvent('fiatOrders', FiatOrderSet.fromArray([]))

    const createSelector = selectors.activityTxs.createFullActivity({
      nftsNetworkNameToAssetName: NFTS_NETWORK_TO_ASSET_NAME,
    })

    const selector = createSelector({
      assetName: 'ethereum',
      walletAccount: 'exodus_0',
    })

    const result = selector(store.getState())

    expect(result.length).toEqual(3)

    const limitedSelector = createSelector({
      assetName: 'ethereum',
      walletAccount: 'exodus_0',
      limit: 1,
    })

    const result2 = limitedSelector(store.getState())

    expect(result2.length).toEqual(1)
  })

  test('should work without personalNotes module', () => {
    const { store, selectors, handleEvent, assets } = setup({ withPersonalNotes: false })
    const fixtures = {
      ethereum: [
        {
          txId: 'tx-id-1',
          error: null,
          date: '2019-07-22T13:54:28.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '-0.23637286 ETH',
          coinName: 'ethereum',
          feeCoinName: 'ethereum',
          feeAmount: '0.000189 ETH',
          to: '0xa96b536eef496e21f5432fd258b6f78cf3673f74',
          currencies: { ethereum: assets.ethereum.currency },
        },
        {
          txId: 'tx-id-2',
          error: null,
          date: '2019-07-22T14:18:09.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '0.24511145 ETH',
          coinName: 'ethereum',
          currencies: { ethereum: assets.ethereum.currency },
        },
        {
          txId: 'tx-id-3',
          error: null,
          date: '2019-07-23T11:23:39.000Z',
          confirmations: 1,
          meta: {},
          token: null,
          dropped: false,
          coinAmount: '0.31622234 ETH',
          coinName: 'ethereum',
          currencies: { ethereum: assets.ethereum.currency },
        },
      ],
    }

    handleEvent('activityTxs', {
      exodus_0: {
        ethereum: fixtures.ethereum.map((tx) => Tx.fromJSON(tx)),
      },
    })
    handleEvent('connectedOrigins', [])
    handleEvent('nftsTxs', {})
    handleEvent('fiatOrders', FiatOrderSet.fromArray([]))

    const createSelector = selectors.activityTxs.createFullActivity({
      nftsNetworkNameToAssetName: NFTS_NETWORK_TO_ASSET_NAME,
    })

    const selector = createSelector({
      assetName: 'ethereum',
      walletAccount: 'exodus_0',
    })

    const result = selector(store.getState())

    expect(result.length).toEqual(3)
  })
})
