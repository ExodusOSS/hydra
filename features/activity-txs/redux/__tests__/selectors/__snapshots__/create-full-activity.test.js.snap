exports[`createFullActivity should add nfts, fiat, connections to props 1`] = `
[
  {
    "assetName": "ethereum",
    "connection": {
      "foo": "bar",
      "origin": "someOrigin",
    },
    "date": 2019-07-22T13:54:28.000Z,
    "failed": false,
    "id": "nft.ethereum.received-tx",
    "nft": {
      "date": 1,
      "from": "ethereumAddress1",
      "network": "ethereum",
      "nftId": "someId1",
      "sent": undefined,
      "to": "ethereumAddress2",
      "txId": "received-tx",
    },
    "pending": false,
    "personalNote": {
      "message": "message",
      "providerData": {
        "origin": "someOrigin",
      },
      "txId": "received-tx",
    },
    "received": true,
    "sent": false,
    "tx": {
      "coinAmount": "1 ETH",
      "coinName": "ethereum",
      "confirmations": 1,
      "currencies": {
        "ethereum": {
          "ETH": 18,
          "Gwei": 9,
          "Kwei": 3,
          "Mwei": 6,
          "finney": 15,
          "szabo": 12,
          "wei": 0,
        },
      },
      "date": "2019-07-22T13:54:28.000Z",
      "dropped": false,
      "feeAmount": "0.000189 ETH",
      "feeCoinName": "ethereum",
      "txId": "received-tx",
      "version": 1,
    },
    "txId": "received-tx",
    "type": "nft",
  },
  {
    "assetName": "ethereum",
    "date": 2019-07-22T13:54:28.000Z,
    "failed": false,
    "fiatOrder": {
      "date": "2023-07-11T15:10:42.226Z",
      "exodusRate": 35,
      "fiatValue": 22,
      "fromAddress": null,
      "fromAmount": 42,
      "fromAsset": "USD",
      "fromWalletAccount": null,
      "orderId": "buy-order",
      "orderType": "buy",
      "provider": "ramp",
      "providerRate": 32,
      "status": "",
      "toAddress": "toAddress",
      "toAmount": 42,
      "toAsset": "ethereum",
      "toWalletAccount": "exodus_0",
      "txId": "buy-order-tx",
    },
    "id": "ethereum.buy-order-tx",
    "pending": false,
    "personalNote": undefined,
    "received": true,
    "sent": false,
    "tx": {
      "coinAmount": "1 ETH",
      "coinName": "ethereum",
      "confirmations": 1,
      "currencies": {
        "ethereum": {
          "ETH": 18,
          "Gwei": 9,
          "Kwei": 3,
          "Mwei": 6,
          "finney": 15,
          "szabo": 12,
          "wei": 0,
        },
      },
      "date": "2019-07-22T13:54:28.000Z",
      "dropped": false,
      "feeAmount": "0.000189 ETH",
      "feeCoinName": "ethereum",
      "txId": "buy-order-tx",
      "version": 1,
    },
    "txId": "buy-order-tx",
    "type": "fiat",
  },
]
`;
